serialize/src/serde.rs

```diff
@@ -0,0 +1,254 @@
+use ark_std::ops::{Deref, DerefMut};
+
+#[cfg(feature = "serde")]
+use ark_std::string::ToString;
+
+use crate::*;
+
+/// A serde-compatible wrapper that forces compression and skips validation.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+pub struct CompressedUnchecked<T>(pub T);
+
+/// A serde-compatible wrapper that skips compression and skips validation.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+pub struct UncompressedUnchecked<T>(pub T);
+
+/// A serde-compatible wrapper that forces compression and validates.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+pub struct CompressedChecked<T>(pub T);
+
+/// A serde-compatible wrapper that skips compression and forces validation.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+pub struct UncompressedChecked<T>(pub T);
+
+macro_rules! impl_ops {
+    ($type:ty, $cons:expr) => {
+        impl<T> Deref for $type {
+            type Target = T;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+
+        impl<T> DerefMut for $type {
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                &mut self.0
+            }
+        }
+
+        impl<T> From<T> for $type {
+            fn from(value: T) -> $type {
+                $cons(value)
+            }
+        }
+
+        impl<T: Valid> Valid for $type {
+            fn check(&self) -> Result<(), SerializationError> {
+                self.0.check()
+            }
+        }
```

Comment:
Probably this should move into `impl_canonical` as it is logically related to `CanonicalDeserialize`